game/yodesk.ksy

```diff
@@ -0,0 +1,1190 @@
+  structures.
+
+  With every new game, Yoda Stories generates a new world. This is done by
+  picking a random sample of puzzles from `PUZ2`. One of the chosen puzzles
+  will be the goal, which when solved wins the game.
+  Each puzzle provides an item when solved and some require one to be completed.
+  During world generation a global world map of 10x10 sectors is filled with
+  zones based on the selected puzzles.
+
+  To add variety and interactivity to each zone the game includes a simple
+  scripting engine. Zones can declare actions that when executed can for
+  example set, move or delete tiles, drop items or activate enemies.
+
+seq:
+  - id: catalog
+    type: catalog_entry
+    repeat: until
+    repeat-until: _.type == "ENDF"
+
+types:
+  catalog_entry:
+    seq:
+      - id: type
+        type: str
+        size: 4
+      - id: size
+        type: u4
+        if: type != "VERS" and type != "ZONE"
+      - id: content
+        type:
+          switch-on: type
+          cases:
+             '"VERS"': version
+             '"STUP"': startup_image
+             '"CHAR"': characters
+             '"CAUX"': character_auxiliaries
+             '"CHWP"': character_weapons
+             '"PUZ2"': puzzles
+             '"SNDS"': sounds
+             '"TILE"': tiles
+             '"TNAM"': tile_names
+             '"ZONE"': zones
+             '"ENDF"': endf
+             _: unknown_catalog_entry
+  unknown_catalog_entry:
+    seq:
+      - id: data
+        size: _parent.size
+  version:
+    seq:
+      - id: version
+        doc: Version of the file. This value is always set to 512.
+        type: u4
+  startup_image:
+    doc: |
+      A 288x288 bitmap to be shown while other assets are loaded and a new
+      world is generated.
+    seq:
+      - id: pixels
+        size: _parent.size
+  sounds:
+    doc: |
+      This section declares sounds used in the game. The actual audio data is
+      stored in wav files on the disk (in a directory named `sfx`) so this
+      section contains paths to each sound file.
+      Sounds can be referenced from the scripting language (see `play_sound`
+      instruction opcode below) and from weapon (see `character` structure).
+      Some sound ids (like the one when the hero is hit, or can't leave a
+      zone) are hard coded in the game.
+    seq:
+      - id: count
+        type: s2
+      - id: sounds
+        type: prefixed_strz
+        repeat: expr
+        repeat-expr: -count
+  tile_names:
+    seq:
+      - id: names
+        doc: |
+          List of tile ids and their corresponding names. These are shown in
+          the inventory or used in dialogs (see `speak_hero` and `speak_npc`
+          opcodes).
+        type: tile_name
+        repeat: until
+        repeat-until:  _.tile_id == 0xFF_FF
+  tile_name:
+    seq:
+      - id: tile_id
+        type: u2
+      - id: name
+        type: strz
+        size: 0x18
+        if: tile_id != 0xFF_FF
```

Comment:
I replaced some occurrences of the pattern with an instance in a6119768d6b45a3f5b3da65f78cc2d295ccb9879. Is there a way to globally define the instance? I don't really think the additional abstraction is justified if only on field uses the check.

Reply:
> I don't really think the additional abstraction is justified if only on field uses the check.

Right, this is focused mainly on places where the check is repeated multiple times, as for example [in the `character` type](https://github.com/kaitai-io/kaitai_struct_formats/blob/2e2b4ea332c84c6ec5c61165cb0218518f867675/game/yodesk.ksy#L1083-L1119). If it only used once (maybe twice), I'm happy with leaving it in `if`.

Reply:
Hmm, I've just realized that this comment of mine is pretty much useless because I suggest to pull the adjacent fields with `if: index != 0xFF_FF` in the [very next comment](https://github.com/kaitai-io/kaitai_struct_formats/pull/403#discussion_r662628427). You've done the `character_body` change too (which is a better way than using the value instance `is_end`) so I'd say feel free to ignore this.